# Veri Biliminin Enstrumanları: Pandas & Numpy

- Numpy nedir? nasıl kullanılır?
- Pandas nedir? nasıl kullanılır?

# Numpy nedir? nasıl kullanılır?

NumPy, Python programlama dili için bir kütüphanedir; büyük, çok boyutlu diziler ve matrisler için destek ekler ve bu dizilerde çalışacak geniş bir üst düzey matematiksel işlev koleksiyonu sunar. Veri bilimi için olmazsa olmazdır

In [1]:
import numpy as np 

Kütüphanemizi ‘np’ ismi ile ‘import’ etme işlemini yaptık. Siz böyle yapmak zorunda değilsiniz. Ama literatür de ‘np’ olarak adlandırılır. Bu en basit işlemi yaptıktan sonra Data oluşturacağız ve bu büyülü dünyaya adım atacağız.

### Numpy’e İlk Adım ‘Arrayler’

Arrayler’i bir çeşit veri depolama olarak görebilirsiniz. Numpy’nin veriyi kullanabilmesi için verileri ‘Array’ haline getirmeliyiz. Array’in ne olduğundan bahsetmedik değil mi ? Arrayler aslında birer matrislerdir. Görebilmek için öncelikle bir veri yaratalım.

In [2]:
Data = [[1,2,3],[4,5,6],[7,8,9]]

Verimizi de yarattığımıza göre ilk Array’i yaratabiliriz.


In [6]:
# Datayı Array formuna çevirebilmek için 
Array = np.array(Data)
Array

array([[1, 2, 3],
       [4, 5, 6],
       [7, 8, 9]])

Buradan sonra bizim ne yapmak istediğimiz önemlidir. Diyelim ki ‘0.’ indekste ki veriyi merak ediyoruz. o halde şöyle yapmalıyız.

In [8]:
Array[0]

array([1, 2, 3])

Matris içinde ki başka bir veriyi öğrenmek istiyoruz ve 2. indeksin 2. verisini öğrenmemiz lazım diyelim o halde şöyle yapacaktık.

In [9]:
Array[2,2]

9

## Numpy’nin Olmazsa Olmaz Fonksiyonları
### Arange Fonksiyonu:
Arange fonksiyonu, verdiğiniz 2 sayı arasını Array’e çevirir.


In [10]:
np.arange(0,100) # bu fonk 0 ila 100 değerleri arasını array olarak depolar

array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16,
       17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33,
       34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50,
       51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67,
       68, 69, 70, 71, 72, 73, 74, 75, 76, 77, 78, 79, 80, 81, 82, 83, 84,
       85, 86, 87, 88, 89, 90, 91, 92, 93, 94, 95, 96, 97, 98, 99])

Bununla kalmıyor, verdiğiniz değer kadar atlayarak da gidebiliyor. Mesela şu örnekte 5'er 5'er atladık.

In [11]:
np.arange(0,100,5)

array([ 0,  5, 10, 15, 20, 25, 30, 35, 40, 45, 50, 55, 60, 65, 70, 75, 80,
       85, 90, 95])

### Zeros, Ones ve Eye Fonksiyonları:
3 Fonksiyonu birlikte aldık çünkü birbirine oldukça yakınlar.
Zeros Fonksiyonu, belirttiğiniz miktar kadar 0'lardan oluşan bir Array yaratacaktır.

In [13]:
np.zeros(100)# 100 adet 0'dan olusan array bastır

array([0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.])

In [14]:
np.ones(100)# 100 adet 1 den olusan array bastır

array([1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
       1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
       1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
       1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
       1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
       1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.])

In [15]:
np.zeros((25,4)) # ones gibi fonksiyonları da bu şekilde yazabilirsiniz

array([[0., 0., 0., 0.],
       [0., 0., 0., 0.],
       [0., 0., 0., 0.],
       [0., 0., 0., 0.],
       [0., 0., 0., 0.],
       [0., 0., 0., 0.],
       [0., 0., 0., 0.],
       [0., 0., 0., 0.],
       [0., 0., 0., 0.],
       [0., 0., 0., 0.],
       [0., 0., 0., 0.],
       [0., 0., 0., 0.],
       [0., 0., 0., 0.],
       [0., 0., 0., 0.],
       [0., 0., 0., 0.],
       [0., 0., 0., 0.],
       [0., 0., 0., 0.],
       [0., 0., 0., 0.],
       [0., 0., 0., 0.],
       [0., 0., 0., 0.],
       [0., 0., 0., 0.],
       [0., 0., 0., 0.],
       [0., 0., 0., 0.],
       [0., 0., 0., 0.],
       [0., 0., 0., 0.]])

Ve bu bölümün sonra son fonksiyonu olan Eye fonksiyonuna gelelim. Kendisini tarif etmek biraz zor. Ama işlevi oldukça basit.

In [16]:
np.eye(10)

array([[1., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 1., 0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 1., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 1., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 1., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 1., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 1., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 1., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 1., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 0., 1.]])

### Linspace Fonskiyonu:
Bu fonksiyon, genel olarak 3 değer alır. İlk 2 değer arasında olan sayıları 3. sayıya eşit olarak dağıtır. Anlamamış olabilirsiniz ama aşağıdaki kod parçasıya kolayca anlayabiliceğinizi düşünüyorum.

In [17]:
np.linspace(0,100,5)# 0 ve 100 arasını 5 eşit parçaya böl ve göster

array([  0.,  25.,  50.,  75., 100.])

### Random Fonksiyonları:
Bu bölümde pek çok fonksiyon göreceksiniz. Hepsinin ayrı bir işlevi var. Ve oldukça önemliler. Zira en fazla kullanılan fonksiyonlar arasında yer alıyorlar.

In [18]:
np.random.randint(0,100)# 0 ve 100 arasında ranstgele bir sayı üret 
# ve ayrıca şu ikisi birbirine eşittir np.random.randint(0,100) = np.random.randint(100)

84

Ayrıca tıpkı diğerleri gibi şu özelliğe de sahiptir.


In [19]:
np.random.randint(0,100,5)# 0 ve 100 arasında 5 tane rastgele sayı bastır

array([55, 78, 51, 72,  2])

Şimdi ki fonskiyonumuz ise 0 ile 1 arasında ki random sayıları verecektir.

In [20]:
np.random.rand(5)# 0 ve 1 arasında 5 değer bastır

array([0.78777872, 0.95106867, 0.42292908, 0.91331576, 0.37813727])

Benzer olup ama aynı olmayan bu fonksiyon ise -1 ile 1 arası sayıları bastırır. Gauss distribution metoduyla yapılmaktadır.


In [21]:
np.random.randn(5) # -1 ve 1 arasında 5 değeri 'gauss distribution' ile bastır

array([1.28882025, 1.16510377, 1.22420842, 0.12284436, 0.26332935])

Şimdi Array’e şekil vermeyi göreceğiz. Matris ile işlemlerimizi icra ettiğimiz için bu fonksiyon ayrı bir işlevseldir. Veriye istediğimiz şekli vermemizi sağlamaktadır.

In [22]:
Data1 = np.arange(225)
Data1

array([  0,   1,   2,   3,   4,   5,   6,   7,   8,   9,  10,  11,  12,
        13,  14,  15,  16,  17,  18,  19,  20,  21,  22,  23,  24,  25,
        26,  27,  28,  29,  30,  31,  32,  33,  34,  35,  36,  37,  38,
        39,  40,  41,  42,  43,  44,  45,  46,  47,  48,  49,  50,  51,
        52,  53,  54,  55,  56,  57,  58,  59,  60,  61,  62,  63,  64,
        65,  66,  67,  68,  69,  70,  71,  72,  73,  74,  75,  76,  77,
        78,  79,  80,  81,  82,  83,  84,  85,  86,  87,  88,  89,  90,
        91,  92,  93,  94,  95,  96,  97,  98,  99, 100, 101, 102, 103,
       104, 105, 106, 107, 108, 109, 110, 111, 112, 113, 114, 115, 116,
       117, 118, 119, 120, 121, 122, 123, 124, 125, 126, 127, 128, 129,
       130, 131, 132, 133, 134, 135, 136, 137, 138, 139, 140, 141, 142,
       143, 144, 145, 146, 147, 148, 149, 150, 151, 152, 153, 154, 155,
       156, 157, 158, 159, 160, 161, 162, 163, 164, 165, 166, 167, 168,
       169, 170, 171, 172, 173, 174, 175, 176, 177, 178, 179, 18

Örneğin burada ‘Arange’ fonksiyonu ile 225'e kadar değer bastırmışız. Ama şekilsiz ve istediğimiz gibi değil.

In [23]:
Data1.reshape(15,15)
#RESHAPE metodu datayı istediğimiz şekli vermemizi sağlar ve matrise çevirir

array([[  0,   1,   2,   3,   4,   5,   6,   7,   8,   9,  10,  11,  12,
         13,  14],
       [ 15,  16,  17,  18,  19,  20,  21,  22,  23,  24,  25,  26,  27,
         28,  29],
       [ 30,  31,  32,  33,  34,  35,  36,  37,  38,  39,  40,  41,  42,
         43,  44],
       [ 45,  46,  47,  48,  49,  50,  51,  52,  53,  54,  55,  56,  57,
         58,  59],
       [ 60,  61,  62,  63,  64,  65,  66,  67,  68,  69,  70,  71,  72,
         73,  74],
       [ 75,  76,  77,  78,  79,  80,  81,  82,  83,  84,  85,  86,  87,
         88,  89],
       [ 90,  91,  92,  93,  94,  95,  96,  97,  98,  99, 100, 101, 102,
        103, 104],
       [105, 106, 107, 108, 109, 110, 111, 112, 113, 114, 115, 116, 117,
        118, 119],
       [120, 121, 122, 123, 124, 125, 126, 127, 128, 129, 130, 131, 132,
        133, 134],
       [135, 136, 137, 138, 139, 140, 141, 142, 143, 144, 145, 146, 147,
        148, 149],
       [150, 151, 152, 153, 154, 155, 156, 157, 158, 159, 160, 161, 162,
       

Ama bu özel metot ile işlemlerimizi çok daha iyi halledebiliyoruz. Ama dikkat etmemiz gereken bir şey var. 225, 15'in karesidir. İlla ki kare sayılar kullanmak zorunda değilsiniz 20 değer verirsiniz ama (5,4)’lük bir matris de yaratabilirsiniz. Şimdi ise kısa öz ve gerekli işlemler yapacağız. Mesela ‘min()’ adı üstünde en küçük değeri verecektir.


In [24]:
Data1.cumsum()# kümülatif olarak toplar

array([    0,     1,     3,     6,    10,    15,    21,    28,    36,
          45,    55,    66,    78,    91,   105,   120,   136,   153,
         171,   190,   210,   231,   253,   276,   300,   325,   351,
         378,   406,   435,   465,   496,   528,   561,   595,   630,
         666,   703,   741,   780,   820,   861,   903,   946,   990,
        1035,  1081,  1128,  1176,  1225,  1275,  1326,  1378,  1431,
        1485,  1540,  1596,  1653,  1711,  1770,  1830,  1891,  1953,
        2016,  2080,  2145,  2211,  2278,  2346,  2415,  2485,  2556,
        2628,  2701,  2775,  2850,  2926,  3003,  3081,  3160,  3240,
        3321,  3403,  3486,  3570,  3655,  3741,  3828,  3916,  4005,
        4095,  4186,  4278,  4371,  4465,  4560,  4656,  4753,  4851,
        4950,  5050,  5151,  5253,  5356,  5460,  5565,  5671,  5778,
        5886,  5995,  6105,  6216,  6328,  6441,  6555,  6670,  6786,
        6903,  7021,  7140,  7260,  7381,  7503,  7626,  7750,  7875,
        8001,  8128,

In [25]:
Data1.min()# en küçük değeri döndürür 

0

In [26]:
Data1.max()# en buyuk değeri döndürür

224

In [27]:
Data1.sum()# tüm veriyi toplar

25200

Şimdi ki fonksiyonumuz ise ‘argmax()’. Ondan kısaca bahsetmek gerekirse en büyük değere sahip değerin (bu veride 224), indexini verecektir (ne yazık ki o da 224).

In [28]:
Data1.argmax()# bu sizi şaşırtmasın.Bunun anlamı en büyük sayının indexini vericektir

224

In [29]:
Data1.argmin()#en küçük sayının indexini vericektir

0

Şimdi ki fonksiyonumuz ise Determinant alabilmek için.

In [30]:
Data2 = np.array([[1,2],[3,4]])
np.linalg.det(Data2)# bu fonksiyon determinantını alır

-2.0000000000000004

In [31]:
stddev = np.std(Data1) # Standart sapmayı bulur.
stddev

64.95126378036586

In [32]:
variance = np.var(Data1)#varyansını bulur
variance

4218.666666666667

### Array Yapılarının İndexlenmesi:


In [34]:
Data1 = np.arange(0,100)
Data1

array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16,
       17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33,
       34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50,
       51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67,
       68, 69, 70, 71, 72, 73, 74, 75, 76, 77, 78, 79, 80, 81, 82, 83, 84,
       85, 86, 87, 88, 89, 90, 91, 92, 93, 94, 95, 96, 97, 98, 99])

In [36]:
Data1[:6]
# 0 dan 6. indexe kadar bastır ayrıca yanda ki kod ile de eşdeğerdir. Data1[0,6]

array([0, 1, 2, 3, 4, 5])

In [37]:
Data1[::3]# baştan sonra 3 atlaya atlaya ilerle

array([ 0,  3,  6,  9, 12, 15, 18, 21, 24, 27, 30, 33, 36, 39, 42, 45, 48,
       51, 54, 57, 60, 63, 66, 69, 72, 75, 78, 81, 84, 87, 90, 93, 96, 99])

Anlamış olduğunuz üzere ‘:’ işareti o index’e dair her veriyi kapsamaktadır. İlk ‘:’ ‘baştan’ anlamına gelirken, son ‘:’ ‘sona doğru anlamına’ gelmektedir.

In [38]:
Data1[:1] = 10 # baştan 2. indexle kadar olanları 10a eşitle
Data1 # ve datamıza baktıgımızda ilk 2 indexin 10 a eşitlendiğini görüyoruz

array([10,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16,
       17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33,
       34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50,
       51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67,
       68, 69, 70, 71, 72, 73, 74, 75, 76, 77, 78, 79, 80, 81, 82, 83, 84,
       85, 86, 87, 88, 89, 90, 91, 92, 93, 94, 95, 96, 97, 98, 99])

Yukarıda ki işlem ise bazen gerekli olabilen bir işlemdir. Veriyi analiz ederken bazen veriler eksik olabiliyor. Bazen 0 veya 1 atamamız gerekebiliyor. O yüzden önemi büyük bir özelliktir.

In [40]:
Data1 > 40 

array([False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True])

Burada gördüğünüz işlemde Data1'de ki veriler eğer 40 değerinden küçük ise ‘False’ değerini, büyükse ‘True’ değerini döndürecektir.

In [41]:
Data1[Data1 > 40]

array([41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57,
       58, 59, 60, 61, 62, 63, 64, 65, 66, 67, 68, 69, 70, 71, 72, 73, 74,
       75, 76, 77, 78, 79, 80, 81, 82, 83, 84, 85, 86, 87, 88, 89, 90, 91,
       92, 93, 94, 95, 96, 97, 98, 99])

### Numpy Temel Operasyonlar:

Genel olarak temel matematiksel işlemler ve diğer alanları kapsamaktadır. İlerledikçe anlayacaksınız. Ama önce yeni bir veri yaratalım

In [42]:
NewArray = np.array([5,10,15,20,25])
NewArray1 = np.array([1,2,3,4,5])

In [43]:
NewArray + NewArray1

array([ 6, 12, 18, 24, 30])

In [44]:
NewArray - NewArray1

array([ 4,  8, 12, 16, 20])

In [45]:
NewArray / NewArray1

array([5., 5., 5., 5., 5.])

In [46]:
NewArray * NewArray1

array([  5,  20,  45,  80, 125])

Şekilde görüldüğü gibi 2 veri arasında 4 işlem ve daha fazlası yapılabilmektedir. Hatta bu verilerle istenilen değer, 4 işlem yada daha fazlası yoluyla kazandırılabilir. Çok daha detaylı bilgiyi buradan araştırabilirsiniz.

### Numpy Transpose Kullanımı:
Matrisin satırlarını sütun, sütunlarını satır yapma işlemi Numpy ile “ .T” yazmak kadar kolay. Aşağıdaki durum herşeyi açıklayabilicek seviyede.;

In [47]:
a = np.array([[5,4]])
a.T

array([[5],
       [4]])

In [49]:
a = [3,6,9]
b = np.array(a)
b.T #a 1 boyutlu olduğu için transpozunu alamadı

array([3, 6, 9])

In [50]:
b = np.array([a])
b.T

array([[3],
       [6],
       [9]])

In [52]:
b = np.array([3,6,9])
b.shape

(3,)

In [53]:
b = np.array([[3,6,9]])
b.shape

(1, 3)

# Pandas nedir? nasıl kullanılır?

Pandas, Python programlama dili için yüksek performanslı, kullanımı kolay veri yapıları ve veri analiz araçları sağlayan açık kaynaklı bir kütüphanedir. ‘.csv’ ve ‘.txt’ dosyalarını açmak ve içerisinde bulunan verileri okuyarak istenen sonuca kolayca ulaşmak için kullanılmaktadır. Yani Pandas sayesinde bir excel dosyasını açarak içerisinde bulunan bir sütunu veya satırı seçip işlem yapabiliriz.

In [54]:
import pandas as pd

### Pandas Serileri

Seriler Numpy dizileri baz alınarak oluşturuldukları için onlara çok benzerler. Seri, etiketli verilerden oluşan tek boyutlu bir veri yapısıdır. Etiket değerlerine ise indeks denir. Verinin kendisi sayılar, dizeler veya başka Python objelerinden oluşabilir. Serileri oluşturmak için ise listeler, sıralı diziler ya da sözlükler kullanılabilir. Pandas Serileri aşağıdaki değişkenleri alırlar

<img src="g41.png">

Burada da ‘Bir seri nasıl oluşturulur?’ sorusunun cevabını görüyorsunuz

In [58]:
Label_list = ['I','am','Learning','Data','Science']
Data_List = [1,2,3,4,5]
Pd_Series1 = pd.Series(Data_List,Label_list)
Pd_Series1

I           1
am          2
Learning    3
Data        4
Science     5
dtype: int64

Ama illaki bu yolla ya da bu uzunlukta olmak zorunda değildir, Şu şekilde de olabilir.

In [57]:
data = np.array(['a','b','c','d'])
Series = pd.Series(data,[100,101,102,103])
Series

100    a
101    b
102    c
103    d
dtype: object

Şimdi 3 tane Seri oluşturalım ve bu seriler üzerinde bir takım işlemler uygulayalım.


In [59]:
DataDict = {'Michael_s exam result': 35, 'Olivia_s exam result': 85}
A = pd.Series(DataDict)
A

Michael_s exam result    35
Olivia_s exam result     85
dtype: int64

In [60]:
DataDict2 = {'Michael_s exam result': 44}
B = pd.Series(DataDict2)
B

Michael_s exam result    44
dtype: int64

In [65]:
DataDict3 = {'Darth_Vader_s exam result' :99}
C = pd.Series(DataDict3)
C

Darth_Vader_s exam result    99
dtype: int64

Görmüş olduğunuz gibi 3 adet seri oluşturduk. Sıra kullanabileceğimiz bazı durumlarda.

In [66]:
A + B

Michael_s exam result    79.0
Olivia_s exam result      NaN
dtype: float64

Bu durumlardan ilki: iki seri arasında dört işlem ve daha fazlası yapılabilir. Ancak dikkat etmeniz gereken bir husus var: ‘A’ değerinin içinde ‘Michael’ ve ‘Olivia’ kişilerinin sınav sonuçları varken, ‘B’ değerinde sadece ‘Michael’ kişisinin sınav sonucu var. İki değeri işleme soktuğumuzda sadece eşleşen değerler işleme tabii olur. Eşleşmeyen değer olursa (Bu örnekte ‘Olivia’ kişisi), o değer ‘NaN’ olarak görünecektir.


‘Peki NaN nedir?’ diye soruyorsanız, ‘Not a number’ın kısaltılmış halidir ve bu veri tipinin sayı olmadığını ifade eder.

In [67]:
DataDict4 = C.append(A)
DataDict4

Darth_Vader_s exam result    99
Michael_s exam result        35
Olivia_s exam result         85
dtype: int64

Yukarıda görüldüğü gibi ‘DataDict4’ diye bir değer oluşturup, ‘C’ değerine ‘A’ değerini ekledik ve üç kişinin de sınav sonuçlarını görebildik. Gerçi ‘Darth Vader’ bu kadar yüksek not almışsa muhtemelen ‘Gücün karanlık tarafı’ ile ilgili bir sınavdır. :)


In [68]:
A['Michael_s exam result'], C['Darth_Vader_s exam result']

(35, 99)

Yukarıda bulunan kodu hemen açıklayayım. ‘A’ değerinde ‘Michael’ ve ‘Olivia’ kişilerinin sınav sonuçları vardı. Diyelim ki biz ‘Michael’ kişisinin sınav sonucunu öğrenmek istiyoruz. Bu durumda ilk kodu yazmamız gerekir. Aynısı ‘Darth Vader’ karakteri için de geçerlidir. Bu illaki sayısal bir değer olmak zorunda değildir (int veya float). ‘String’ yani kelimelerden oluşan bir veri de olabilirdi.

### Pandas DataFrame

DataFrameler, Türkçesiyle ‘Veri Çerçeveleri’ Pandas kütüphanesinde asıl olayın döndüğü kısımdır ve pek çok işlemi bu kısımda yapacağız. Burada sütunlar ‘Column’ ya da ‘Feature’ olarak satırlar ise ‘row’ ya da ‘ indeks’ olarak adlandırılır.
Öncelikle kendimize bir veri oluşturmalıyız.

In [111]:
from numpy.random import randn

df = pd.DataFrame(data = randn(5,5), index = ['A','B','C','D','E'], columns = ['Columns1','Columns2','Columns3','Columns4','Columns5'])
df

Columns1  Columns2  Columns3  Columns4  Columns5
A -1.787872  0.101395 -0.186640  1.390045 -0.294629
B -1.116155  0.209617  0.919080 -1.501275  1.012764
C -0.123932 -0.208232 -0.182487  0.036994  0.679899
D -2.172706  0.251268 -0.558812 -1.567580 -1.204402
E  1.999596  0.411939 -1.948896 -0.911635  2.274610

Çalıştığınız veride size gerekli olan sütunları şu yöntemle alabilirsiniz.

In [112]:
df['Columns1']

A   -1.787872
B   -1.116155
C   -0.123932
D   -2.172706
E    1.999596
Name: Columns1, dtype: float64

Üstelik sadece 1 sütun değil aynı yöntemle birden fazla sütunu da alabiliyorsunuz.


In [113]:
df[['Columns1','Columns5']]

Columns1  Columns5
A -1.787872 -0.294629
B -1.116155  1.012764
C -0.123932  0.679899
D -2.172706 -1.204402
E  1.999596  2.274610

Yeni bir sütun ekleme işlemini ise şöyle yapıyoruz.


In [114]:
df['Columns6'] = pd.Series(randn(5),['A','B','C','D','E'])
df

Columns1  Columns2  Columns3  Columns4  Columns5  Columns6
A -1.787872  0.101395 -0.186640  1.390045 -0.294629 -0.874086
B -1.116155  0.209617  0.919080 -1.501275  1.012764  1.760548
C -0.123932 -0.208232 -0.182487  0.036994  0.679899 -2.602868
D -2.172706  0.251268 -0.558812 -1.567580 -1.204402  0.139146
E  1.999596  0.411939 -1.948896 -0.911635  2.274610 -0.467513

Bununla da sınırlı değil, istediğiniz sütuna ya da satıra istediğiniz işlemi de uygulayabilirsiniz.

In [115]:
df['Columns7'] = (df['Columns6'] + df['Columns4'] - df['Columns1'] ) / df['Columns2'] * df['Columns3']
df

Columns1  Columns2  Columns3  Columns4  Columns5  Columns6   Columns7
A -1.787872  0.101395 -0.186640  1.390045 -0.294629 -0.874086  -4.240706
B -1.116155  0.209617  0.919080 -1.501275  1.012764  1.760548   6.030649
C -0.123932 -0.208232 -0.182487  0.036994  0.679899 -2.602868  -2.140025
D -2.172706  0.251268 -0.558812 -1.567580 -1.204402  0.139146  -1.655234
E  1.999596  0.411939 -1.948896 -0.911635  2.274610 -0.467513  15.984929

Örneğin, yukarıda ‘Column7’ değerini oluşturup onu çeşitli sütunların işlemler sonucu oluşturduğu bir değere eşitledik ve böyle bir tablo oluştu.

Peki ‘Column’ ekleriz de çıkaramaz mıyız ?

In [116]:
df.drop('Columns2', axis = 1, inplace = True)

‘Axis’ in ne olduğunu merak edenler için kendisinin ‘default’ değeri 0'dır ve ‘0’ satırları ‘1’ sütunlarını simgeler. Buradaki ‘inplace’ parametresi önemlidir, bu işlemin kalıcı olup olmadığını belirler ve ‘True’ dediğimiz zaman kalıcı olarak atama yapar.

In [117]:
df.set_index('Column ismi', inplace = True)

KeyError: "None of ['Column ismi'] are in the columns"

Aşğıdaki şekilde de ‘index’ ve ‘Column’ların isimlerini öğrenebiliyoruz.

In [118]:
df.index

Index(['A', 'B', 'C', 'D', 'E'], dtype='object')

In [119]:
df.columns

Index(['Columns1', 'Columns3', 'Columns4', 'Columns5', 'Columns6', 'Columns7'], dtype='object')

Şimdi ise ‘loc’ ve ‘iloc’ durumunu işleyelim.


In [120]:
df.loc['C']

Columns1   -0.123932
Columns3   -0.182487
Columns4    0.036994
Columns5    0.679899
Columns6   -2.602868
Columns7   -2.140025
Name: C, dtype: float64

Bu kodu kullandığımızda, bize ‘C’ satırında ki ‘Column’ değerlerini vermektedir.

In [121]:
df.loc['A']

Columns1   -1.787872
Columns3   -0.186640
Columns4    1.390045
Columns5   -0.294629
Columns6   -0.874086
Columns7   -4.240706
Name: A, dtype: float64

In [122]:
df.iloc[0]

Columns1   -1.787872
Columns3   -0.186640
Columns4    1.390045
Columns5   -0.294629
Columns6   -0.874086
Columns7   -4.240706
Name: A, dtype: float64

Bu ikisi arasındaki bağlantıyı oturup bir 10 saniye düşünmenizi istiyorum. 

Şunu farkedeceksiniz ki ikisi de oldukça benzer şeyler.
‘iloc’ fonksiyonu, satırın indeksine göre ‘Column’ değerlerini çıkartır ve satırların indeksleri bildiğiniz gibi 0'dan başlayarak ilerliyor. ‘A’nın indeksi 0, ve ‘iloc’ fonksiyonuna ‘0’ verilmiş. Bu durumda bu iki fonksiyon aslında aynı şeyi temsil ediyor. ‘loc’ fonksiyonunda isim belirtmeniz gerekirken, ‘iloc’ da indeks belirtmeniz gerekiyor.

Bazen daha ileri gitmek gerekir, mesela aşağıdaki gibi.

In [123]:
df.loc['A','Columns5']

-0.2946292613015262

### DataFrame Filtreleme İşlemleri

In [124]:
df

Columns1  Columns3  Columns4  Columns5  Columns6   Columns7
A -1.787872 -0.186640  1.390045 -0.294629 -0.874086  -4.240706
B -1.116155  0.919080 -1.501275  1.012764  1.760548   6.030649
C -0.123932 -0.182487  0.036994  0.679899 -2.602868  -2.140025
D -2.172706 -0.558812 -1.567580 -1.204402  0.139146  -1.655234
E  1.999596 -1.948896 -0.911635  2.274610 -0.467513  15.984929

In [125]:
df > 0.2

Columns1  Columns3  Columns4  Columns5  Columns6  Columns7
A     False     False      True     False     False     False
B     False      True     False      True      True      True
C     False     False     False      True     False     False
D     False     False     False     False     False     False
E      True     False     False      True     False      True

In [126]:
booleanDF = df > 0

‘Boolean’ı şöyle açıklayabiliriz. İstenilen değer (burada ‘0’dan büyük olma şartı ile), koşulu sağlıyorsa ‘True’ değilse ‘False’ değerini döndürür.

In [127]:
df[booleanDF]

Columns1  Columns3  Columns4  Columns5  Columns6   Columns7
A       NaN       NaN  1.390045       NaN       NaN        NaN
B       NaN   0.91908       NaN  1.012764  1.760548   6.030649
C       NaN       NaN  0.036994  0.679899       NaN        NaN
D       NaN       NaN       NaN       NaN  0.139146        NaN
E  1.999596       NaN       NaN  2.274610       NaN  15.984929

Burada ise ‘NaN’ değerleri üstteki Boolean değerindeki ‘False’ değerlerini temsil etmekte.

In [128]:
df[df > 0.5]

Columns1  Columns3  Columns4  Columns5  Columns6   Columns7
A       NaN       NaN  1.390045       NaN       NaN        NaN
B       NaN   0.91908       NaN  1.012764  1.760548   6.030649
C       NaN       NaN       NaN  0.679899       NaN        NaN
D       NaN       NaN       NaN       NaN       NaN        NaN
E  1.999596       NaN       NaN  2.274610       NaN  15.984929

Burada ise ‘df’ içinde ‘0.5’ten büyük olanları gösterir. ‘False’ olanlar yine ‘NaN’ olarak gösteriliyor. Bunu daha ileride daha iyi örneklerle açıklayacağız.

In [129]:
df['Columns1'] > 0

A    False
B    False
C    False
D    False
E     True
Name: Columns1, dtype: bool

Yukarıda ise belli bir ‘Column’ değerinin filtrelemesini görüyorsunuz. Aşağıda ise bunu her ‘Column’ değerine uygulanışını görüyorsunuz. Bazılarında sonuç farklı. Bazılarında 4 satır varken bazılarında 2 satır var, sizce neden olabilir ?

In [130]:
df[df['Columns3']> 0]

Columns1  Columns3  Columns4  Columns5  Columns6  Columns7
B -1.116155   0.91908 -1.501275  1.012764  1.760548  6.030649

In [131]:
df[df['Columns4']> 0]

Columns1  Columns3  Columns4  Columns5  Columns6  Columns7
A -1.787872 -0.186640  1.390045 -0.294629 -0.874086 -4.240706
C -0.123932 -0.182487  0.036994  0.679899 -2.602868 -2.140025

In [132]:
df[df['Columns5']> 0]

Columns1  Columns3  Columns4  Columns5  Columns6   Columns7
B -1.116155  0.919080 -1.501275  1.012764  1.760548   6.030649
C -0.123932 -0.182487  0.036994  0.679899 -2.602868  -2.140025
E  1.999596 -1.948896 -0.911635  2.274610 -0.467513  15.984929

In [133]:
df[df['Columns1']> 0]

Columns1  Columns3  Columns4  Columns5  Columns6   Columns7
E  1.999596 -1.948896 -0.911635   2.27461 -0.467513  15.984929

Kısaca özetlemek gerekirse, ‘True’ olan değerler gösterilir. Örneğin, ‘Column4’te 4 satır bulunmakta demek ki 5. satır ‘True’ değerinde.

In [134]:
df[(df['Columns1']> 0) & (df['Columns3']> 0)]

Empty DataFrame
Columns: [Columns1, Columns3, Columns4, Columns5, Columns6, Columns7]
Index: []

In [135]:
df[(df['Columns1']> 0) | (df['Columns5']> 1)]

Columns1  Columns3  Columns4  Columns5  Columns6   Columns7
B -1.116155  0.919080 -1.501275  1.012764  1.760548   6.030649
E  1.999596 -1.948896 -0.911635  2.274610 -0.467513  15.984929

Yukarıda çeşitli filtreleme alıştırmalarını görüyorsunuz. Önceden bahsettiğimiz kurallara tabiidirler. ‘&’ işareti ‘and’ anlamında ve ‘|’işareti ‘or’ anlamındadır. Bu işaretler yerine ‘and’ ya da ‘or’ kullanır iseniz hata alırsınız. ‘&’ işareti ‘SHIFT’ ve ‘6’ tuşları ile ‘|’ işareti ise ‘ALT GR’ ve ‘-’ ile yapılmaktadır.

In [136]:
df['Columns6'] = ['NewValue1','NewValue2','NewValue3','NewValue4','NewValue5']
df

Columns1  Columns3  Columns4  Columns5   Columns6   Columns7
A -1.787872 -0.186640  1.390045 -0.294629  NewValue1  -4.240706
B -1.116155  0.919080 -1.501275  1.012764  NewValue2   6.030649
C -0.123932 -0.182487  0.036994  0.679899  NewValue3  -2.140025
D -2.172706 -0.558812 -1.567580 -1.204402  NewValue4  -1.655234
E  1.999596 -1.948896 -0.911635  2.274610  NewValue5  15.984929

Verimizi güncellerken 2. fotoğrafta anlaşıldığı üzere verilen değer ‘String’ olsa bile direk atandığıdır.

In [137]:
df.set_index('Columns6' , inplace = True)
df

Columns1  Columns3  Columns4  Columns5   Columns7
Columns6                                                    
NewValue1 -1.787872 -0.186640  1.390045 -0.294629  -4.240706
NewValue2 -1.116155  0.919080 -1.501275  1.012764   6.030649
NewValue3 -0.123932 -0.182487  0.036994  0.679899  -2.140025
NewValue4 -2.172706 -0.558812 -1.567580 -1.204402  -1.655234
NewValue5  1.999596 -1.948896 -0.911635  2.274610  15.984929

Yukarıda ilk satırda kalıcı bir şekilde index ismini ‘Column6’ olarak atadık. 2. satırda index başlığını sorguladık.

## GroupBy Operasyonları

GroupBy Operasyonları, Sql tablolarındankiyle tamamen aynı. Sql bilmiyor iseniz bile rahat olun. Zor bir konu değil ancak önemli bir konu. İstenilen durumu gruplar gösterir.

In [138]:
data = {'Job': ['Data Mining','CEO','Lawyer','Lawyer','Data Mining','CEO'],'Labouring': ['Immanuel','Jeff','Olivia','Maria','Walker','Obi-Wan'], 'Salary': [4500,30000,6000,5250,5000,35000]}
data

{'Job': ['Data Mining', 'CEO', 'Lawyer', 'Lawyer', 'Data Mining', 'CEO'],
 'Labouring': ['Immanuel', 'Jeff', 'Olivia', 'Maria', 'Walker', 'Obi-Wan'],
 'Salary': [4500, 30000, 6000, 5250, 5000, 35000]}

Verimizi oluşturduktan sonra, DataFrame’e atıyoruz.

In [139]:
df = pd.DataFrame(data)
df

Job Labouring  Salary
0  Data Mining  Immanuel    4500
1          CEO      Jeff   30000
2       Lawyer    Olivia    6000
3       Lawyer     Maria    5250
4  Data Mining    Walker    5000
5          CEO   Obi-Wan   35000

In [145]:
SalaryGroupBy = df.groupby('Salary')

Ve ‘Salary’ özelliğine göre gruplamış olduk.

In [146]:
SalaryGroupBy.sum()

Job Labouring
Salary                       
4500    Data Mining  Immanuel
5000    Data Mining    Walker
5250         Lawyer     Maria
6000         Lawyer    Olivia
30000           CEO      Jeff
35000           CEO   Obi-Wan

In [143]:
SalaryGroupBy.min()

Job Labouring
Salary                       
4500    Data Mining  Immanuel
5000    Data Mining    Walker
5250         Lawyer     Maria
6000         Lawyer    Olivia
30000           CEO      Jeff
35000           CEO   Obi-Wan

In [144]:
SalaryGroupBy.max()

Job Labouring
Salary                       
4500    Data Mining  Immanuel
5000    Data Mining    Walker
5250         Lawyer     Maria
6000         Lawyer    Olivia
30000           CEO      Jeff
35000           CEO   Obi-Wan

Gördüğünüz gibi hangi fonksiyonu kullanırsanız kullanın, tablo halinde yansıtacaktır.

In [147]:
df.groupby('Salary').sum()# Bu şekilde de işlem kısaltılabilir.

Job Labouring
Salary                       
4500    Data Mining  Immanuel
5000    Data Mining    Walker
5250         Lawyer     Maria
6000         Lawyer    Olivia
30000           CEO      Jeff
35000           CEO   Obi-Wan

Son iki fotoğraftaki işlemi yukarıdaki kod tek başına yapmaktadır.

In [148]:
df.groupby('Job').sum().loc['CEO']

Salary    65000
Name: CEO, dtype: int64

‘CEO’ların maaşların toplamı.

In [149]:
df.groupby('Job').count()

Labouring  Salary
Job                           
CEO                  2       2
Data Mining          2       2
Lawyer               2       2

In [150]:
df.groupby('Job').min()

Labouring  Salary
Job                          
CEO              Jeff   30000
Data Mining  Immanuel    4500
Lawyer          Maria    5250

Numpy kütüphanesindeki fonksiyonlar burada da kullanılabilmektedir.

In [151]:
df.groupby('Job').min()['Salary']

Job
CEO            30000
Data Mining     4500
Lawyer          5250
Name: Salary, dtype: int64

In [152]:
df.groupby('Job').min()['Salary']['Lawyer']

5250

In [153]:
df.groupby('Job').mean()['Salary']['CEO']

32500

Burada ise, mesleğe göre sıralayıp Ceo’ların ortalama maaşını bulduk. GroupBy Operasyonları, görüldüğü gibi zor bir konu değil. Oldukça kısa ve öz bir konu.

## Concatenate Merge Ve Join Fonksiyonları

İlk olarak Concatenate ile başlayalım. Temel olarak birleştirme işlemini bu fonksiyon ile yapmaktayız. Aynı listelerdeki ‘zip’ fonksiyonu gibidir.


In [155]:
data = {'A': ['A1','A2','A3','A4'],'B': ['B1','B2','B3','B4'],'C': ['C1','C2','C3','C4']}
data1 = {'A': ['A5','A6','A7','A8'],'B': ['B5','B6','B7','B8'],'C': ['C5','C6','C7','C8']}
df1 = pd.DataFrame(data, index = [1,2,3,4])
df2 = pd.DataFrame(data1, index = [5,6,7,8])
df1

A   B   C
1  A1  B1  C1
2  A2  B2  C2
3  A3  B3  C3
4  A4  B4  C4

In [156]:
df2

A   B   C
5  A5  B5  C5
6  A6  B6  C6
7  A7  B7  C7
8  A8  B8  C8

İki tane veri seti oluşturalım. Ve aşağıdaki gibi birleştirelim.

In [157]:
pd.concat([df1,df2])

A   B   C
1  A1  B1  C1
2  A2  B2  C2
3  A3  B3  C3
4  A4  B4  C4
5  A5  B5  C5
6  A6  B6  C6
7  A7  B7  C7
8  A8  B8  C8

Şöyle de birleştirilebilirdi.

In [158]:
pd.concat([df1,df2], axis = 1)

A    B    C    A    B    C
1   A1   B1   C1  NaN  NaN  NaN
2   A2   B2   C2  NaN  NaN  NaN
3   A3   B3   C3  NaN  NaN  NaN
4   A4   B4   C4  NaN  NaN  NaN
5  NaN  NaN  NaN   A5   B5   C5
6  NaN  NaN  NaN   A6   B6   C6
7  NaN  NaN  NaN   A7   B7   C7
8  NaN  NaN  NaN   A8   B8   C8

Burada olmayan veriler elbetteki ‘NaN’ olarak atanıyor. Bu işlem yalnızca 2 değil daha fazla sayıda olan durumlarda da kullanılabilir.


<img src="g42.png">

Join, ile devam edelim. ‘Join’ fonkiyonları adı üstünde birbirlerine katılma işlemleri olarak düşünülebilir. Ve lisedeki ‘kümeler’ konusu olarak da düşünülebilir. Öyleyse iki veri oluşturup bu işlemleri yapmaya başlayalım.


In [159]:
data = {'A': ['A1','A2','A3','A4'],'B': ['B1','B2','B3','B4'],'C': ['C1','C2','C3','C4']}
data1 = {'A': ['A5','A6','A7','A8'],'B': ['B5','B6','B7','B8'],'C': ['C5','C6','C7','C8']}
df1 = pd.DataFrame(data, index = [1,2,3,4])
df2 = pd.DataFrame(data1, index = [5,6,7,8])
df1

A   B   C
1  A1  B1  C1
2  A2  B2  C2
3  A3  B3  C3
4  A4  B4  C4

In [166]:
data2 = {'id': [1,2,6,7,8], 'Feature1': ['K','M','O','Q','S'], 'Feature2': ['L', 'N', 'P', 'R', 'T']}
df2 = pd.DataFrame(data2, columns = ['id', 'Feature1', 'Feature2'])
df2

id Feature1 Feature2
0   1        K        L
1   2        M        N
2   6        O        P
3   7        Q        R
4   8        S        T

‘Left Join’ ile başlayalım.

<img src="g43.png">

Left Join, Sql’deki ‘Left Join’ ile birebir aynıdır. İşlevi ise yukarıda verilmiştir. Siz burada ‘df1'i table1 olarak, ‘df2’yi table2 olarak düşünmelisiniz.

In [167]:
df1.join(df2)

A   B   C  id Feature1 Feature2
1  A1  B1  C1   2        M        N
2  A2  B2  C2   6        O        P
3  A3  B3  C3   7        Q        R
4  A4  B4  C4   8        S        T

In [169]:
df2.join(df1)

id Feature1 Feature2    A    B    C
0   1        K        L  NaN  NaN  NaN
1   2        M        N   A1   B1   C1
2   6        O        P   A2   B2   C2
3   7        Q        R   A3   B3   C3
4   8        S        T   A4   B4   C4

Right Join, ile de aynı mantık yürütülebilir. İşlevi ve kodu aşağıda verilmiştir.

In [170]:
df1.join(df2, how = 'right')

A    B    C  id Feature1 Feature2
0  NaN  NaN  NaN   1        K        L
1   A1   B1   C1   2        M        N
2   A2   B2   C2   6        O        P
3   A3   B3   C3   7        Q        R
4   A4   B4   C4   8        S        T

<img src="g44.png">
<img src="g45.png">

‘Full Outer Joın’ yazdığına bakmayın. O Sql’deki ismi, burada sadece ‘outer’ olarak geçiyor ve aşağıda ikisinin de örnekleri var.

In [171]:
df1.join(df2, how ='outer')

A    B    C  id Feature1 Feature2
0  NaN  NaN  NaN   1        K        L
1   A1   B1   C1   2        M        N
2   A2   B2   C2   6        O        P
3   A3   B3   C3   7        Q        R
4   A4   B4   C4   8        S        T

In [172]:
df1.join(df2, how = 'inner')

A   B   C  id Feature1 Feature2
1  A1  B1  C1   2        M        N
2  A2  B2  C2   6        O        P
3  A3  B3  C3   7        Q        R
4  A4  B4  C4   8        S        T

In [176]:
frames = [df1,df2]
df_keys = pd.concat(frames, keys=['x', 'y'])
df_keys

A    B    C   id Feature1 Feature2
x 1   A1   B1   C1  NaN      NaN      NaN
  2   A2   B2   C2  NaN      NaN      NaN
  3   A3   B3   C3  NaN      NaN      NaN
  4   A4   B4   C4  NaN      NaN      NaN
y 0  NaN  NaN  NaN  1.0        K        L
  1  NaN  NaN  NaN  2.0        M        N
  2  NaN  NaN  NaN  6.0        O        P
  3  NaN  NaN  NaN  7.0        Q        R
  4  NaN  NaN  NaN  8.0        S        T

### Merge İşlemleri

In [177]:
dataset1 = {'A':['A1','A2','A3'], 'B': ['B1','B2','B3'], 'Key': ['K1','K2','K3']}
dataset2 = {'X':['X1','X2','X3','X4'], 'Y': ['Y1','Y2','Y3','Y4'], 'Key': ['K1','K2','K3','K4']}
df1 = pd.DataFrame(dataset1,index = [1,2,3])
df2 = pd.DataFrame(dataset2,index = [1,2,3,4])
df1

A   B Key
1  A1  B1  K1
2  A2  B2  K2
3  A3  B3  K3

In [178]:
df2

X   Y Key
1  X1  Y1  K1
2  X2  Y2  K2
3  X3  Y3  K3
4  X4  Y4  K4

In [179]:
pd.merge(df1,df2, on = 'Key')

A   B Key   X   Y
0  A1  B1  K1  X1  Y1
1  A2  B2  K2  X2  Y2
2  A3  B3  K3  X3  Y3

Ve ilk parametremizle tanıştık bile. ‘On’ parametresine anahtarı yazmanız gerekmektedir. Anahtar sizin verinizde önemli yeri olmalıdır ki gerçek manada ‘anahtar’ olabilsin.


In [180]:
pd.merge(df1,df2, on = 'Key', how = 'left')

A   B Key   X   Y
0  A1  B1  K1  X1  Y1
1  A2  B2  K2  X2  Y2
2  A3  B3  K3  X3  Y3

‘How’ parametresine zaten Join konusundan alışığız. Orada gördüğünüz ‘inner’, ‘outer’ , ‘right’ değerleri burada da geçerli.

In [181]:
pd.merge(df1,df2, on = 'Key', how = 'right')

A    B Key   X   Y
0   A1   B1  K1  X1  Y1
1   A2   B2  K2  X2  Y2
2   A3   B3  K3  X3  Y3
3  NaN  NaN  K4  X4  Y4

In [182]:
pd.merge(df1,df2, on = 'Key', how = 'outer')

A    B Key   X   Y
0   A1   B1  K1  X1  Y1
1   A2   B2  K2  X2  Y2
2   A3   B3  K3  X3  Y3
3  NaN  NaN  K4  X4  Y4

In [183]:
pd.merge(df1,df2, on = 'Key', how = 'inner')

A   B Key   X   Y
0  A1  B1  K1  X1  Y1
1  A2  B2  K2  X2  Y2
2  A3  B3  K3  X3  Y3

In [185]:
pd.merge(df1,df2, left_index=True, right_index=True, how='outer')

A    B Key_x   X   Y Key_y
1   A1   B1    K1  X1  Y1    K1
2   A2   B2    K2  X2  Y2    K2
3   A3   B3    K3  X3  Y3    K3
4  NaN  NaN   NaN  X4  Y4    K4

## DataFrame Operasyonları Ve Pivot Table
Burada küçük ama işlevsel fonksiyonları gördükten sonra ‘Pivot Table’ konusunu işleyeceğiz. Önce verimizi oluşturalım ve başlayalım

In [186]:
data = {'Column1': [1,2,3,4,5,6], 'Column2': [1000,1000,2000,3000,3000,1000],'Column3': ['Mace Windu','Darth Vader','Palpatine','Kylo Ren','Rey','Obi-Wan']}
df = pd.DataFrame(data)
df

Column1  Column2      Column3
0        1     1000   Mace Windu
1        2     1000  Darth Vader
2        3     2000    Palpatine
3        4     3000     Kylo Ren
4        5     3000          Rey
5        6     1000      Obi-Wan

Ve küçük ama işlevsel fonksiyonlarımıza gelelim.

In [187]:
df.head()

Column1  Column2      Column3
0        1     1000   Mace Windu
1        2     1000  Darth Vader
2        3     2000    Palpatine
3        4     3000     Kylo Ren
4        5     3000          Rey

In [188]:
df.head(n = 2) # df.head(2) = df.head(n = 2)

Column1  Column2      Column3
0        1     1000   Mace Windu
1        2     1000  Darth Vader

‘head()’ ilk 5 değeri göstermekte ama içine aldığı değer kadar da gösterebilir. Örnekte 2 verilmiş. Aynı şekilde ‘tail()’de son 5 değeri gösterir. O da ‘head()’ gibi değer alabilir.

In [189]:
df.describe()

Column1      Column2
count  6.000000     6.000000
mean   3.500000  1833.333333
std    1.870829   983.192080
min    1.000000  1000.000000
25%    2.250000  1000.000000
50%    3.500000  1500.000000
75%    4.750000  2750.000000
max    6.000000  3000.000000

In [190]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6 entries, 0 to 5
Data columns (total 3 columns):
 #   Column   Non-Null Count  Dtype 
---  ------   --------------  ----- 
 0   Column1  6 non-null      int64 
 1   Column2  6 non-null      int64 
 2   Column3  6 non-null      object
dtypes: int64(2), object(1)
memory usage: 272.0+ bytes


In [191]:
df['Column2'].unique()

array([1000, 2000, 3000], dtype=int64)

In [192]:
df['Column2'].nunique()

3

‘unique()’ verilen verinin kaç adet ‘eşşiz’ verisi olduğunu bize verir. Aşağıdaki kod ise kaç adet ‘eşşiz’ değer olduğunu veriyor.

In [193]:
df['Column2'].value_counts()

1000    3
3000    2
2000    1
Name: Column2, dtype: int64

‘value_counts()’ ise hangi değerden kaç adet olduğunu veren bir fonksiyon.

In [194]:
df[df['Column1'] >= 2]

Column1  Column2      Column3
1        2     1000  Darth Vader
2        3     2000    Palpatine
3        4     3000     Kylo Ren
4        5     3000          Rey
5        6     1000      Obi-Wan

In [195]:
df[(df['Column1'] >= 2) & (df['Column2'] == 300) ]

Empty DataFrame
Columns: [Column1, Column2, Column3]
Index: []

Anlaşıldığı üzere ‘DataFrame Filtreleme’ işlemlerini de burada kullanabiliyoruz.

In [196]:
def square(x): 
    return x ** 2
square(2)

4

Burada bir Fonksiyon oluşturup içine aldığı değerin karesini vermesini sağladık.

In [198]:
df['Column2']

0    1000
1    1000
2    2000
3    3000
4    3000
5    1000
Name: Column2, dtype: int64

In [197]:
df['Column2'].apply(square)

0    1000000
1    1000000
2    4000000
3    9000000
4    9000000
5    1000000
Name: Column2, dtype: int64

Bu fonksiyonu ‘apply()’ ile verimize attığımızda bize karelerini verdi.

Bu özel durumu projelerinizde kullanabilirsiniz. Örneğin, aracın motor kısmının verisini inceliyorsunuz ve ekibiniz sizden motora ait bir bulgu istedi. Onlardan gerekli olan matematiksel formülü alıp aynı bu örnekteki gibi kullanıp rahatlıkla işinizi yapabilirsiniz.

Ama bunun daha kısa bir yolu var.

In [199]:
df['Column2'].apply(lambda x : x **2 )

0    1000000
1    1000000
2    4000000
3    9000000
4    9000000
5    1000000
Name: Column2, dtype: int64

In [201]:
df['Column2'] = df['Column2'].apply(square) # Güncelleyebilmek için bunu kullanıyoruz. Ayrıca lambdayıda kullanabirdik.

Ama bu işlemleri yapınca veriniz otomatik olarak güncellenmiyor. Sizin güncellemeniz gerekiyor, onu da yukarıdaki gibi yapıyoruz.

In [202]:
df['Column3'].apply(len) # String'in uzunluğunu alma

0    10
1    11
2     9
3     8
4     3
5     7
Name: Column3, dtype: int64

In [203]:
df.drop('Column3', axis = 1 )

Column1  Column2
0        1  1000000
1        2  1000000
2        3  4000000
3        4  9000000
4        5  9000000
5        6  1000000

Şimdi ise ‘sort_values()’e geçelim.

Aşağıda aldığı parametreleri görüyorsunuz. ‘by’ neye göre alınacağını belirtirken, ‘ascending’ sıralamasını neye göre olacağını belirtir. Küçükten büyüğe ya da büyükten küçüğe olarak sıralayabilirsiniz. ‘Axis’ ve ‘İnplace’ parametrelerini zaten biliyorsunuz. ‘kind’ parametresi aşağıda yazan 3 adete göre verinizi düzenleyebilirsiniz. Çok kullanılmaz ama isterseniz araştırabilirsiniz. ‘na_postion’ ise ‘NaN’ değerlerinin pozisyonunu belirtmenizi sağlar.

<img src="g46.png">

In [204]:
df

Column1  Column2      Column3
0        1  1000000   Mace Windu
1        2  1000000  Darth Vader
2        3  4000000    Palpatine
3        4  9000000     Kylo Ren
4        5  9000000          Rey
5        6  1000000      Obi-Wan

In [205]:
df.sort_values(by=['Column1', 'Column3']) # Birden fazla değerde 'Sort' edilebilir.

Column1  Column2      Column3
0        1  1000000   Mace Windu
1        2  1000000  Darth Vader
2        3  4000000    Palpatine
3        4  9000000     Kylo Ren
4        5  9000000          Rey
5        6  1000000      Obi-Wan

Bu fonksiyon belli bir değere göre sıralamamızı sağlar. Biz burada ‘Column2’ye göre sıralamasını sağladık. Yukarı da ise ‘Column1’ ve ‘Column3’e sıraladık ki farkı görebilesiniz.

In [206]:
df.sort_values(by=['Column1', 'Column2'])

Column1  Column2      Column3
0        1  1000000   Mace Windu
1        2  1000000  Darth Vader
2        3  4000000    Palpatine
3        4  9000000     Kylo Ren
4        5  9000000          Rey
5        6  1000000      Obi-Wan

In [207]:
df.sort_values('Column2', ascending = True) # Küçükten Büyüğe

Column1  Column2      Column3
0        1  1000000   Mace Windu
1        2  1000000  Darth Vader
5        6  1000000      Obi-Wan
2        3  4000000    Palpatine
3        4  9000000     Kylo Ren
4        5  9000000          Rey

In [208]:
df.sort_values('Column2', ascending = False) # Büyükten küçüğe;

Column1  Column2      Column3
3        4  9000000     Kylo Ren
4        5  9000000          Rey
2        3  4000000    Palpatine
0        1  1000000   Mace Windu
1        2  1000000  Darth Vader
5        6  1000000      Obi-Wan

In [209]:
df.sort_values('Column1', kind = 'heapsort')

Column1  Column2      Column3
0        1  1000000   Mace Windu
1        2  1000000  Darth Vader
2        3  4000000    Palpatine
3        4  9000000     Kylo Ren
4        5  9000000          Rey
5        6  1000000      Obi-Wan

In [210]:
df.sort_values('Column1', kind = 'mergesort')

Column1  Column2      Column3
0        1  1000000   Mace Windu
1        2  1000000  Darth Vader
2        3  4000000    Palpatine
3        4  9000000     Kylo Ren
4        5  9000000          Rey
5        6  1000000      Obi-Wan

In [211]:
df.sort_values('Column1', kind = 'quicksort')

Column1  Column2      Column3
0        1  1000000   Mace Windu
1        2  1000000  Darth Vader
2        3  4000000    Palpatine
3        4  9000000     Kylo Ren
4        5  9000000          Rey
5        6  1000000      Obi-Wan

Gelelim ‘Pivot Table’ konusuna. Excel biliyorsanız korkacak bir şey yok. Çünkü tamamıyla aynı. Bilmiyorsanız da endişelenmeyin. Çünkü çok basit bir kullanımı var.

In [214]:
df = pd.DataFrame({'Month': ['January','February','March','January','February','March','January','February','March'],'State':['New York','New York','New York','Texas','Texas','Texas','Washington','Washington','Washington'],
'moisture': [20,25,65,34,56,85,21,56,79]})
df

Month       State  moisture
0   January    New York        20
1  February    New York        25
2     March    New York        65
3   January       Texas        34
4  February       Texas        56
5     March       Texas        85
6   January  Washington        21
7  February  Washington        56
8     March  Washington        79

In [215]:
df.pivot_table(index = 'Month', columns = 'State', values = 'moisture')

State     New York  Texas  Washington
Month                                
February        25     56          56
January         20     34          21
March           65     85          79

In [216]:
df.pivot_table(index = 'State', columns = 'Month', values = 'moisture')

Month       February  January  March
State                               
New York          25       20     65
Texas             56       34     85
Washington        56       21     79

## Veri Setini Okuma Yöntemleri

In [ ]:
dataset = pd.read_csv('Dosya_Yolu\Data_Name.csv') # Pc'de IDE'nizde bu şekilde okuyabilirsiniz.

excelset = pd.read_excel('Excels_Name.xlsx') # Excel Dosyasını okumak için
excelset.to_excel('excelnewfile.xlsx') # Excel dosyasına çevirme

New = pd.read_html(' Datasetin urlsi.html') # İnternette ki bir veriyi okumak için